### PR TITLE
Migrate to CRDs - WIP

### PR DESCRIPTION
**Pull request by enxebre**
Migrate to CRDs:
- [x] New api generation and layout based on kubebuilder
-- kubebuilder init --domain k8s.io --license apache2 --owner “The Kubernetes Authors”
-- kubebuilder create api --group awsproviderconfig --version v1alpha1 --kind AWSMachineProviderConfig
- [x] Move the cloud/ code from root folder into `pkg/cloud`
- [x] revendor
- [x] Fix machine status, provider status
- [x] Update `tests/` and `cmd/aws-actuator`
- [x] Update docker file and Make

How to run it:
```kustomize build vendor/sigs.k8s.io/cluster-api/config/default/ >> cluster-api-bootstrap.yaml```
```kubectl apply -f cluster-api-bootstrap.yaml```
```make manager```
```./bin/manager```

How to regenerate code:
```go generate ./pkg/... ./cmd/...```

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+	// Setup all Controllers
+	if err := controller.AddToManager(mgr); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("Starting the Cmd.")
+
+	// Start the Cmd
+	log.Fatal(mgr.Start(signals.SetupSignalHandler()))
```
Comment (frobware) [inline cmd/manager/main.go]: If `SIGTERM` is received does Start() return? If so, I wouldn't expect that to print a fatal error.

File: pkg/apis/addtoscheme_awsproviderconfig_v1alpha1.go [hunk 2]
```diff
@@ -14,21 +14,13 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package options
+package apis
```
Comment (frobware) [inline pkg/apis/addtoscheme_awsproviderconfig_v1alpha1.go]: This looks like it was in the right package `options`, but is now in `apis`.

Comment (frobware) [reply pkg/apis/addtoscheme_awsproviderconfig_v1alpha1.go]: > This looks like it was in the right package `options`, but is now in `apis`.

Once again, the GH review layout confuses me.

**Comment by ingvagabund** [comment]
https://github.com/kubernetes-sigs/cluster-api/pull/546 is definitely blocker

File: pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go [hunk 3]
```diff
@@ -21,6 +21,15 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Annotation constants
+const (
+	// ClusterNameLabel is the label that a machineset must have to identify the
+	// cluster to which it belongs.
+	ClusterNameLabel = "sigs.k8s.io/cluster-api-cluster"
```
[the remark this answers is not in the document]
Comment (enxebre) [reply pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go]: this is a carry over, there's bunch of things to consolidate after this. Let's address any non specific to CRDs move in different follow ups

**Pull request merged**